titus-server-master/src/main/java/com/netflix/titus/master/eviction/service/TaskTerminationExecutor.java

```diff
@@ -0,0 +1,132 @@
+        this.jobOperations = jobOperations;
+        this.quotasManager = quotasManager;
+        this.serializedInvoker = ReactorSerializedInvoker.<Void>newBuilder()
+                .withName("taskTerminationExecutor")
+                .withMaxQueueSize(MAX_QUEUE_SIZE)
+                .withExcessiveRunningTime(TASK_EXCESSIVE_RUNNING_TIMEOUT)
+                .withScheduler(scheduler)
+                .withClock(titusRuntime.getClock())
+                .withRegistry(titusRuntime.getRegistry())
+                .build();
+    }
+
+    void shutdown() {
+        serializedInvoker.shutdown(Duration.ofSeconds(30));
+    }
+
+    Flux<EvictionEvent> events() {
+        return eventProcessor;
+    }
+
+    public Mono<Void> terminateTask(String taskId, String reason) {
+        return Mono
+                .defer(() -> {
+                    Pair<Job<?>, Task> jobTaskPair = checkTaskIsRunningOrThrowAnException(taskId);
+                    Job<?> job = jobTaskPair.getLeft();
+                    Task task = jobTaskPair.getRight();
+
+                    if (getQuota(job) <= 0) {
+                        return Mono.error(newEvictionRejectionEvent(taskId, reason, job));
+                    }
+
+                    return serializedInvoker.submit(Mono.defer(() -> {
+                        if (!quotasManager.tryConsumeQuota(job, task)) {
+                            return Mono.error(newEvictionRejectionEvent(taskId, reason, job));
+                        }
+                        return ReactorExt.toMono(jobOperations.killTask(taskId, false, reason)).timeout(TASK_TERMINATE_TIMEOUT);
+                    }));
+                })
+                .doFinally(signalType -> {
+                    if (signalType == SignalType.ON_COMPLETE) {
+                        eventProcessor.onNext(EvictionEvent.newTaskTerminationEvent(taskId, reason, true));
+                    }
+                });
+    }
+
+    private Pair<Job<?>, Task> checkTaskIsRunningOrThrowAnException(String taskId) {
+        Optional<Pair<Job<?>, Task>> jobAndTask = jobOperations.findTaskById(taskId);
+        if (!jobAndTask.isPresent()) {
+            throw EvictionException.taskNotFound(taskId);
+        }
+        Task task = jobAndTask.get().getRight();
+        TaskState state = task.getStatus().getState();
+        if (state == TaskState.Accepted) {
+            throw EvictionException.taskNotScheduledYet(task);
+        }
+        if (!TaskState.isBefore(state, TaskState.KillInitiated)) {
+            throw EvictionException.taskAlreadyStopped(task);
+        }
+        return jobAndTask.get();
+    }
+
+    private Long getQuota(Job<?> job) {
+        return quotasManager.findJobEvictionQuota(job.getId()).map(EvictionQuota::getQuota).orElse(0L);
+    }
+
+    private EvictionException newEvictionRejectionEvent(String taskId, String reason, Job<?> job) {
+        String constraintReason = quotasManager.explainJobQuotaConstraints(job.getId()).orElse("Job no longer around");
```

Comment:
`Job no longer around` -> `Job does not exist`

Reply:
Changed